activedirectory.sgnl.yaml

```diff
@@ -0,0 +1,390 @@
+description: "Active Directory as a System of Record"
+hostname: "{{Input Required: Hostname}}"
+defaultSyncFrequency: HOURLY
+defaultSyncMinInterval: 1
+defaultApiCallFrequency: SECONDLY
+defaultApiCallMinInterval: 1
+type: "ActiveDirectory-1.0.0"
+# Example Config:
+# {
+#   "baseDN":"dc=org,dc=example,dc=io",
+#   "certificateChain": ". . . base64 encoded certificate chain here . . "
+#   "entityConfig": {
+#     "User": {
+#       "selector": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*))"
+#     },
+#     "Computer": {
+#       "selector": "(&(objectCategory=computer)(name=*))"
+#     },
+#     "Group": {
+#       "selector": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*))"
+#     },
+#     "GroupMember": {
+#       "memberOf": "Group",
+#       "selector": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
+#       "memberUniqueIdAttribute": "memberObjectGUID",
+#       "memberOfUniqueIdAttribute": "groupObjectGUID"
+#     }
+#   },
+# }
+adapterConfig: "ewogICJiYXNlRE4iOiAiZGM9Y29ycCxkYz1leGFtcGxlLGRjPWFpIiwKICAiZW50aXR5Q29uZmlnIjogewogICAgIlVzZXIiOiB7CiAgICAgICJzZWxlY3RvciI6ICIoJihvYmplY3RDYXRlZ29yeT11c2VyKShvYmplY3RDbGFzcz11c2VyKShkaXN0aW5ndWlzaGVkTmFtZT0qKSkiCiAgICB9LAogICAgIkNvbXB1dGVyIjogewogICAgICAic2VsZWN0b3IiOiAiKCYob2JqZWN0Q2F0ZWdvcnk9Y29tcHV0ZXIpKG5hbWU9KikpIgogICAgfSwKICAgICJHcm91cCI6IHsKICAgICAgInNlbGVjdG9yIjogIigmKG9iamVjdENhdGVnb3J5PWdyb3VwKShvYmplY3RDbGFzcz1ncm91cCkoZGlzdGluZ3Vpc2hlZE5hbWU9KikpIgogICAgfSwKICAgICJHcm91cE1lbWJlciI6IHsKICAgICAgIm1lbWJlck9mIjogIkdyb3VwIiwKICAgICAgInNlbGVjdG9yIjogIigmKG1lbWJlck9mPXt7Q29sbGVjdGlvbklkfX0pKG9iamVjdENhdGVnb3J5PXVzZXIpKG9iamVjdENsYXNzPXVzZXIpKSIsCiAgICAgICJtZW1iZXJVbmlxdWVJZEF0dHJpYnV0ZSI6ICJtZW1iZXJPYmplY3RHVUlEIiwKICAgICAgIm1lbWJlck9mVW5pcXVlSWRBdHRyaWJ1dGUiOiAiZ3JvdXBPYmplY3RHVUlEIgogICAgfQogIH0sCiAgImNlcnRpZmljYXRlQ2hhaW4iOiAiLiAuIC4gYmFzZTY0IGVuY29kZWQgY2VydGlmaWNhdGUgY2hhaW4gaGVyZSAuIC4gIgp9"
```

Comment:
As mentioned in [previous comments](https://github.com/SGNL-ai/catalog-sgnl-templates/pull/56#discussion_r1588406111), please add correct placeholders for `baseDN`, `certificateChain` etc.

Reply:
```
{
    "baseDN": "{{Input Required: BaseDN}}",
    "entityConfig": {
        "User": {
            "selector": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*))"
        },
        "Computer": {
            "selector": "(&(objectCategory=computer)(name=*))"
        },
        "Group": {
            "selector": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "selector": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
            "memberUniqueIdAttribute": "{{Input Required: If memberOf is present}}",
            "memberOfUniqueIdAttribute": "{{Input Required: If memberOf is present}}"
        }
    },
    "certificateChain": "{{Input Required: If LDAPS is enabled: base64 encoded certificateChain}}"
}
```

Updated default config.

Reply:
Can we replace `selector` here with `query` -- we should seem semantically familiar to folks who are administrators of AD and LDAP systems, and in these fields you enter an LDAP Query

Reply:
Tangentially, @r-rakshith @Mohitp98  I worry we've preemptively generalized here. For Active Directory, we know what `memberUniqueIdAttribute` and `memberOfUniqueIdAttribute` are going to be. At a minimum, we should pre-populate these for an SoR Admin with the appropriate values for AD, but we should chat about further iterations where we remove this setting completely and instead replace it with some higher level setting that defines the type of system we're connecting to, e.g. `"Type": "ActiveDirectory"`. Things like AD are going to have a range of specificities that I don't think we should force an administrator to configure and manage.

Reply:
That sounds good to me. I was thinking of the same given the oddities in Active Directory. We might end up having multiple implementations of the adapter although they're technically LDAP. We should chat more about this!


In the meantime, 
This is what the current configuration example looks like. The initial idea was to use a combination of "selector" to select an entity and "filter" to get only a subset of those.
I have no objections reducing all of that to "query" if that is more suitable for administrators. Let me know what you think/
```json
{
    "baseDN": "dc=org,dc=example,dc=io",
    "certificateChain": "....",
    "entityConfig": {
        "User": {
            "selector": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*))",
            "filter": "(&(logonCount=0))"
        },
        "Group": {
            "selector": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*))",
            "filter": "(&(instanceType=4)(sAMAccountName=Blocked Sign-In))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "selector": "(&(memberOf={{CollectionId}}))",
			"filter": "(&(logonCount=0))",
			"memberUniqueIdAttribute": "memberDistinguishedName",
			"memberOfUniqueIdAttribute": "groupDistinguishedName"
        }
    }
}
```

Reply:
@marc-sgnl , @r-rakshith  Here is the updated config:

Reduced `filter` & `selector` attributes to a single `query` attribute.

```json
{
    "baseDN": "{{Input Required: BaseDN}}",
    "certificateChain": "{{Input Required: If LDAPS is enabled: base64 encoded certificateChain}}",
    "entityConfig": {
        "User": {
            "query": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*)(logonCount=0))"
        },
        "Computer": {
            "query": "(&(objectCategory=computer)(name=*))"
        },
        "Group": {
            "query": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*)(instanceType=4))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "query": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
            "memberUniqueIdAttribute": "{{Input Required: If memberOf is present}}",
            "memberOfUniqueIdAttribute": "{{Input Required: If memberOf is present}}"
        }
    }
}
```